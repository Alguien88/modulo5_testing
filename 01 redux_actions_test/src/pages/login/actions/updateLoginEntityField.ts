import { FieldValidationResult } from 'lc-form-validation';
import { actionIds } from './actionIds';
import { LoginEntity } from '../viewModel';
import { validations } from '../validations';

export const updateLoginEntityField = (
    loginEntity: LoginEntity,
    fieldName: string,
    value: string
) => ((dispatch) => {
    const promise = validations
        .validateField(loginEntity, fieldName, value)
        .then((validationResult) => {
            const action = updateLoginEntityFieldCompleted(fieldName, value, validationResult);
            if (validationResult.succeeded) {
                dispatch(action);
            }
        });
    return promise;
});

export const updateLoginEntityFieldCompleted = (
    fieldName: string,
    value: string,
    validationResult: FieldValidationResult
) => ({
    type: actionIds.UPDATE_LOGIN_ENTITY_FIELD,
    payload: {
        fieldName,
        value,
        validationResult,
    },
});
