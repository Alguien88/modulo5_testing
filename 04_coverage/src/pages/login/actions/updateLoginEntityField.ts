import { FieldValidationResult } from 'lc-form-validation';
import { actionIds } from './actionIds';
import { validations } from '../validations';
import { LoginEntity } from '../viewModel';

export const updateLoginEntityField = (
  loginEntity: LoginEntity,
  fieldName: string,
  value: string
) => ((dispatch) => {
  const promise = validations
      .validateField(loginEntity, fieldName, value)
      .then((validationResult) => {
          const action = updateLoginEntityFieldCompleted(fieldName, value, validationResult);
          if (validationResult.succeeded) {
              dispatch(action);
          }
      });
  return promise;
});

export const updateLoginEntityFieldCompleted =
  (fieldName: string, value: any, fieldValidationResult: FieldValidationResult) => ({
    type: actionIds.UPDATE_LOGIN_ENTITY_FIELD,
    payload: {
      fieldName,
      value,
      fieldValidationResult,
    },
  });
