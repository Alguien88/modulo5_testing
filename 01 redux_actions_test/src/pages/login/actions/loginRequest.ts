import { LoginEntity } from '../../../rest-api/model/loginEntity';
import { validations } from '../validations';
import { mapLoginEntityVMToModel } from '../mappers';
import { login } from '../../../rest-api/api/login';
import { history } from '../../../history';
import { routes } from '../../../common/constants/routes';
import { actionIds } from './actionIds';

export const loginRequest = (loginEntity: LoginEntity) => (dispatch) => {
    const promise = validations.validateForm(loginEntity).then(
        (validationResult) => {
            const action = displayErrors(validationResult.fieldErrors);
            validationResult.succeeded ?
                doLogin(loginEntity) :
                dispatch(action);
        }
    );
    return promise;
};

const doLogin = (loginEntity: LoginEntity) => {
    const loginEntityModel = mapLoginEntityVMToModel(loginEntity);
    login(loginEntityModel)
        .then(() => history.push(routes.members))
        .catch((error) => console.log(error));
};

const displayErrors = (fieldErrors) => ({
    type: actionIds.UPDATE_LOGIN_FORM_ERRORS,
    payload: fieldErrors,
});
