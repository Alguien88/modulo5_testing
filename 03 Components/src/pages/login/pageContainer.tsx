import { connect } from 'react-redux';
import { State } from '../reducers';
import { LoginPage } from './page';
import { loginRequest } from './actions/loginRequest';
import { updateLoginEntityField } from './actions/updateLoginEntityField';

const mapStateToProps = (state: State) => ({
  loginEntity: state.login.loginEntity,
  loginFormErrors: state.login.loginFormErrors,
});

const mapDispatchToProps = (dispatch) => ({
  updateField: (loginEntity) => (fieldName, value) => dispatch(updateLoginEntityField(loginEntity, fieldName, value)),
  doLogin: (loginEntity) => () => dispatch(loginRequest(loginEntity)),
});

const mergeProps = (stateProps, dispatchProps, ownProps) => ({
  ...ownProps,
  ...stateProps,
  updateField: dispatchProps.updateField(stateProps.loginEntity),
  doLogin: dispatchProps.doLogin(stateProps.loginEnity),
});

export const LoginPageContainer = connect(mapStateToProps, mapDispatchToProps, mergeProps)(LoginPage);
