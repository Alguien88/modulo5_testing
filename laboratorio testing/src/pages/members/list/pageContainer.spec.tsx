import * as React from 'react';
import { shallow } from 'enzyme';
import { MemberListPageContainer } from './pageContainer';
import configureStore from 'redux-mock-store';
import { State } from '../../reducers';
import * as actions from './actions/fetchMembers';
import * as mappers from './mappers';

const getMockStore = configureStore();

describe('pages/members/list/pageContainer tests', () => {
  it('should render as expected passing state', () => {
    // Arrange
    const state = {
      members: [
        {
          id: 1,
          login: 'login 1',
          avatar_url: 'avatar 1',
        },
        {
          id: 2,
          login: 'login 2',
          avatar_url: 'avatar 2',
        },
      ],
    } as State;

    const store = getMockStore(state);

    // Act
    const component = shallow(
      <MemberListPageContainer />, {
        context: { store },
      }
    );

    // Assert
    expect(component).toMatchSnapshot();
  });

  it('should call to fetchMembers action creator when call to fetchMembers prop', () => {
    // Arrange
    const state = {
      members: [
        {
          id: 1,
          login: 'login 1',
          avatar_url: 'avatar 1',
        },
        {
          id: 2,
          login: 'login 2',
          avatar_url: 'avatar 2',
        },
      ],
    } as State;

    const stub = jest.spyOn(actions, 'fetchMembers')
      .mockReturnValue({
        type: 'test type',
      });

    const store = getMockStore(state);

    // Act
    const component = shallow(
      <MemberListPageContainer
      />,
      {
        context: { store },
      }
    );

    component.prop('fetchMembers')('test fieldName', 'test value');

    // Assert
    expect(stub).toHaveBeenCalled();
  });

  it('should call to mapMemberListModelToVM when render component', () => {
    // Arrange
    const state = {
      members: [
        {
          id: 1,
          login: 'login 1',
          avatar_url: 'avatar 1',
        },
        {
          id: 2,
          login: 'login 2',
          avatar_url: 'avatar 2',
        },
      ],
    } as State;

    const store = getMockStore(state);

    const mapMemberListModelToVMStub = jest.spyOn(mappers, 'mapMemberListModelToVM')
      .mockReturnValue({
        type: 'test type',
      });

    // Act
    const component = shallow(
      <MemberListPageContainer
      />,
      {
        context: { store },
      }
    );

    // Assert
    expect(mapMemberListModelToVMStub).toHaveBeenCalledWith(state.members);
  });
});
