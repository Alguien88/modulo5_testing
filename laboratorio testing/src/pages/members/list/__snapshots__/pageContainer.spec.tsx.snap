// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/members/list/pageContainer tests should render as expected passing state 1`] = `
<PageContainer
  fetchMembers={[Function]}
  members={
    Array [
      Object {
        "avatarUrl": "avatar 1",
        "id": 1,
        "name": "login 1",
      },
      Object {
        "avatarUrl": "avatar 2",
        "id": 2,
        "name": "login 2",
      },
    ]
  }
/>
`;
